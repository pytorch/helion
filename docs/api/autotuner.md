# Autotuner Module

The `helion.autotuner` module provides automatic optimization of kernel configurations.

```{eval-rst}
.. currentmodule:: helion.autotuner

.. automodule:: helion.autotuner
   :members:
   :undoc-members:
   :show-inheritance:
```

## Configuration Classes

### Config

```{eval-rst}
.. autoclass:: helion.runtime.config.Config
   :members:
   :undoc-members:
```

## Search Algorithms

The autotuner supports multiple search strategies:

### Differential Evolution

```{eval-rst}
.. automodule:: helion.autotuner.differential_evolution
   :members:
```

### Random Search

```{eval-rst}
.. automodule:: helion.autotuner.random_search
   :members:
```

### Finite Search

```{eval-rst}
.. automodule:: helion.autotuner.finite_search
   :members:
```
