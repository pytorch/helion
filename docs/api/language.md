# Language Module

The `helion.language` module contains the core DSL constructs for writing GPU kernels.

## Loop Constructs

### tile()

```{eval-rst}
.. currentmodule:: helion.language

.. autofunction:: tile
```

The `tile()` function is the primary way to create parallel loops in Helion kernels. It provides several key features:

**Tiling Strategies**: The exact tiling strategy is determined by a Config object, typically created through autotuning. This allows for:
- Multidimensional tiling
- Index swizzling for cache locality
- Dimension reordering
- Flattening of iteration spaces

**Usage Patterns**:

```python
# Simple 1D tiling
for tile in hl.tile(1000):
    # tile.begin, tile.end, tile.block_size are available
    # Load entire tile (not just first element)
    data = tensor[tile]  # or hl.load(tensor, tile) for explicit loading
```

```python
# 2D tiling
for tile_i, tile_j in hl.tile([height, width]):
    # Each tile represents a portion of the 2D space
    pass
```

```python
# With explicit begin/end/block_size
for tile in hl.tile(0, 1000, block_size=64):
    pass
```

**Grid vs Loop Behavior**:
- When used at the top level of a kernel function, `tile()` becomes the grid of the kernel (parallel blocks)
- When used nested inside another loop, it becomes a sequential loop within each block

### grid()

```{eval-rst}
.. autofunction:: grid
```

The `grid()` function iterates over individual indices rather than tiles. It's equivalent to `tile(size, block_size=1)` but returns scalar indices instead of tile objects.

### static_range()

```{eval-rst}
.. autofunction:: static_range
```

`static_range()` behaves like a compile-time unrolled range for small loops. It hints the compiler to fully unroll the loop body where profitable.

## Memory Operations

### load()

```{eval-rst}
.. autofunction:: load
```

### store()

```{eval-rst}
.. autofunction:: store
```

### atomic_add()

```{eval-rst}
.. autofunction:: atomic_add
```

### atomic_and()

```{eval-rst}
.. autofunction:: atomic_and
```

### atomic_or()

```{eval-rst}
.. autofunction:: atomic_or
```

### atomic_xor()

```{eval-rst}
.. autofunction:: atomic_xor
```

### atomic_xchg()

```{eval-rst}
.. autofunction:: atomic_xchg
```

### atomic_max()

```{eval-rst}
.. autofunction:: atomic_max
```

### atomic_min()

```{eval-rst}
.. autofunction:: atomic_min
```

### atomic_cas()

```{eval-rst}
.. autofunction:: atomic_cas
```

## Inline Assembly

### inline_asm_elementwise()

```{eval-rst}
.. autofunction:: inline_asm_elementwise
```

Executes target-specific inline assembly on elements of one or more tensors with broadcasting and optional packed processing.

### inline_triton()

```{eval-rst}
.. autofunction:: inline_triton
```

Embeds small Triton code snippets directly inside a Helion kernel. Common indentation is removed automatically, placeholders are replaced using ``str.format`` with tuple or dict arguments, and the final line in the snippet becomes the return value. Provide tensors (or tuples of tensors) via ``output_like`` so Helion knows the type of the return value.

## Tensor Creation

### zeros()

```{eval-rst}
.. autofunction:: zeros
```

### full()

```{eval-rst}
.. autofunction:: full
```

### arange()

See {func}`~helion.language.arange` for details.

## Tunable Parameters

### register_block_size()

```{eval-rst}
.. autofunction:: register_block_size
```

### register_tunable()

```{eval-rst}
.. autofunction:: register_tunable
```

## Tile Operations

### Tile Class

```{eval-rst}
.. autoclass:: Tile
   :members:
   :undoc-members:
```

The `Tile` class represents a portion of an iteration space with the following key attributes:
- `begin`: Starting indices of the tile
- `end`: Ending indices of the tile
- `block_size`: Size of the tile in each dimension

## View Operations

### subscript()

```{eval-rst}
.. autofunction:: subscript
```

### split()

```{eval-rst}
.. autofunction:: split
```

### join()

```{eval-rst}
.. autofunction:: join
```

## StackTensor
### StackTensor class
```{eval-rst}
.. autoclass:: StackTensor
   :undoc-members:
```

### stacktensor_like
```{eval-rst}
.. autofunction:: stacktensor_like
```

## Reduction Operations

### reduce()

See {func}`~helion.language.reduce` for details.

## Scan Operations

### associative_scan()

See {func}`~helion.language.associative_scan` for details.

### cumsum()

See {func}`~helion.language.cumsum` for details.

### cumprod()

See {func}`~helion.language.cumprod` for details.

### tile_index()

```{eval-rst}
.. autofunction:: tile_index
```

### tile_begin()

```{eval-rst}
.. autofunction:: tile_begin
```

### tile_end()

```{eval-rst}
.. autofunction:: tile_end
```

### tile_block_size()

```{eval-rst}
.. autofunction:: tile_block_size
```

### tile_id()

```{eval-rst}
.. autofunction:: tile_id
```

## Synchronization


### signal()

```{eval-rst}
.. autofunction:: signal
```

### wait()

```{eval-rst}
.. autofunction:: wait
```

## Utilities

### device_print()

See {func}`~helion.language.device_print` for details.

## Constexpr Operations

### constexpr()

See {class}`~helion.language.constexpr` for details.

### specialize()

See {func}`~helion.language.specialize` for details.

## Matrix Operations

### dot()

See {func}`~helion.language.dot` for details.
