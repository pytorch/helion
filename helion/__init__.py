from __future__ import annotations

from triton import cdiv
from triton import next_power_of_2

from . import _compat as _compat_module  # noqa: F401  # side-effect import
from . import _logging
from . import exc
from . import language
from . import runtime
from .runtime import Config
from .runtime import Kernel
from .runtime import kernel
from .runtime import kernel as jit  # alias
from .runtime.settings import RefMode
from .runtime.settings import Settings

__all__ = [
    "Config",
    "Kernel",
    "RefMode",
    "Settings",
    "cdiv",
    "exc",
    "jit",
    "kernel",
    "language",
    "next_power_of_2",
    "runtime",
]

_logging.init_logs()
