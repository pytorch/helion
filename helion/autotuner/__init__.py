from __future__ import annotations

from .config_fragment import BooleanFragment as BooleanFragment
from .config_fragment import EnumFragment as EnumFragment
from .config_fragment import IntegerFragment as IntegerFragment
from .config_fragment import ListOf as ListOf
from .config_fragment import PowerOfTwoFragment as PowerOfTwoFragment
from .config_spec import ConfigSpec as ConfigSpec
from .differential_evolution import (
    DifferentialEvolutionSearch as DifferentialEvolutionSearch,
)
from .finite_search import FiniteSearch as FiniteSearch
from .local_cache import LocalAutotuneCache as LocalAutotuneCache
from .local_cache import StrictLocalAutotuneCache as StrictLocalAutotuneCache
from .pattern_search import PatternSearch as PatternSearch
from .random_search import RandomSearch as RandomSearch

search_algorithms = {
    "DifferentialEvolutionSearch": DifferentialEvolutionSearch,
    "FiniteSearch": FiniteSearch,
    "PatternSearch": PatternSearch,
    "RandomSearch": RandomSearch,
}
