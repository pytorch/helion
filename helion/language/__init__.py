from __future__ import annotations

from .constexpr import ConstExpr as constexpr  # noqa: F401
from .constexpr import specialize as specialize
from .creation_ops import arange as arange
from .creation_ops import full as full
from .creation_ops import zeros as zeros
from .device_print import device_print as device_print
from .inline_asm_ops import inline_asm_elementwise as inline_asm_elementwise
from .loops import grid as grid
from .loops import static_range as static_range
from .loops import tile as tile
from .matmul_ops import dot as dot
from .memory_ops import atomic_add as atomic_add
from .memory_ops import load as load
from .memory_ops import store as store
from .reduce_ops import reduce as reduce
from .scan_ops import associative_scan as associative_scan
from .scan_ops import cumprod as cumprod
from .scan_ops import cumsum as cumsum
from .signal_wait import signal as signal
from .signal_wait import wait as wait
from .slice_proxy import make_slice as make_slice
from .stack_tensor import StackTensor as StackTensor
from .stack_tensor import stacktensor_like as stacktensor_like
from .tile_ops import tile_begin as tile_begin
from .tile_ops import tile_block_size as tile_block_size
from .tile_ops import tile_end as tile_end
from .tile_ops import tile_id as tile_id
from .tile_ops import tile_index as tile_index
from .tile_proxy import Tile as Tile
from .tunable_ops import register_block_size as register_block_size
from .tunable_ops import register_reduction_dim as register_reduction_dim
from .tunable_ops import register_tunable as register_tunable
from .view_ops import subscript as subscript

_MEMORY_OPS = (store, load, atomic_add, wait, signal)
