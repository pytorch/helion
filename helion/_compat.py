from __future__ import annotations

import contextlib
import functools
from typing import Any
from typing import Callable
from typing import cast

import torch
from torch._inductor.runtime.hints import DeviceProperties
from torch._inductor.utils import triton_type
import triton
from triton.backends.compiler import BaseBackend
from triton.backends.compiler import GPUTarget
import triton.language as tl
import triton.runtime.jit as triton_jit

NativeSpecializeImpl = Callable[
    [type[BaseBackend], object, bool, bool, bool], tuple[object, ...]
]
CreateSpecializeImpl = Callable[
    [Callable[..., object]], Callable[..., tuple[object, ...]]
]


def _make_specialize_impl_wrapper(
    *,
    native_impl: NativeSpecializeImpl | None = None,
    create_factory: CreateSpecializeImpl | None = None,
) -> Callable[..., object]:
    if native_impl is None:
        native_impl = cast(
            "NativeSpecializeImpl | None",
            getattr(triton_jit, "native_specialize_impl", None),
        )
    if native_impl is None and create_factory is None:
        raise AttributeError("native_specialize_impl unavailable")

    def specialize_impl_wrapper(
        *args: object,
        **kwargs: object,
    ) -> object:
        specialize_extra = cast(
            "Callable[..., object] | None",
            kwargs.pop("specialize_extra", None),
        )
        kwargs.pop("specialize_zero_one", None)
        backend_param = kwargs.pop("backend", None)
        args_list: list[object] = list(args)
        backend_type: type[BaseBackend]
        if backend_param is None and args_list:
            first = args_list[0]
            if isinstance(first, type) and issubclass(first, BaseBackend):
                backend_type = first
                args_list.pop(0)
            elif isinstance(first, BaseBackend):
                backend_type = type(first)
                args_list.pop(0)
            else:
                backend_type = BaseBackend
        elif isinstance(backend_param, type) and issubclass(backend_param, BaseBackend):
            backend_type = backend_param
        elif isinstance(backend_param, BaseBackend):
            backend_type = type(backend_param)
        else:
            backend_type = BaseBackend

        arg = kwargs.pop("arg", None)
        if arg is None:
            if args_list:
                arg = args_list.pop(0)
            else:
                raise TypeError("specialize_impl() missing positional argument 'arg'")

        def _pop_flag(
            key: str,
            *,
            alt_keys: tuple[str, ...] = (),
            default: bool | None = None,
        ) -> bool:
            value = kwargs.pop(key, None)
            if value is None:
                for alt in alt_keys:
                    value = kwargs.pop(alt, None)
                    if value is not None:
                        break
            if value is None:
                if args_list:
                    value = args_list.pop(0)
                elif default is not None:
                    value = default
                else:
                    raise TypeError(f"specialize_impl() missing argument '{key}'")
            return bool(value)

        is_const = _pop_flag("is_const")
        specialize_value = _pop_flag(
            "specialize_value",
            alt_keys=("specialize",),
            default=True,
        )
        align = _pop_flag("align", default=True)

        if native_impl is not None:
            result = native_impl(
                backend_type,
                arg,
                is_const,
                specialize_value,
                align,
            )
            if specialize_extra is not None:
                with contextlib.suppress(Exception):
                    specialize_extra(arg)
        else:
            assert create_factory is not None

            def _call_specialize_extra(
                extra_arg: object,
                kind: object,
                *,
                align: bool = True,
            ) -> object:
                if specialize_extra is None:
                    return None
                try:
                    return specialize_extra(extra_arg)
                except TypeError:
                    try:
                        return specialize_extra(extra_arg, kind, align=align)
                    except Exception:
                        return None
                except Exception:
                    return None

            impl = create_factory(_call_specialize_extra)
            result = impl(
                arg,
                is_const=is_const,
                specialize_value=specialize_value,
                align=align,
            )
        return result

    return specialize_impl_wrapper


def _ensure_triton_specialize_impl_alias() -> None:
    if hasattr(triton_jit, "specialize_impl"):
        return
    if hasattr(triton_jit, "native_specialize_impl"):
        module: Any = triton_jit
        module.specialize_impl = _make_specialize_impl_wrapper()  # type: ignore[assignment]
        return
    if hasattr(triton_jit, "create_specialize_impl"):
        module: Any = triton_jit
        module.specialize_impl = _make_specialize_impl_wrapper(
            create_factory=triton_jit.create_specialize_impl,
        )  # type: ignore[assignment]


_ensure_triton_specialize_impl_alias()


def _ensure_backend_specialization_alias() -> None:
    if hasattr(BaseBackend, "get_arg_specialization"):
        return
    if hasattr(BaseBackend, "get_tensor_specialization"):
        BaseBackend.get_arg_specialization = BaseBackend.get_tensor_specialization  # type: ignore[attr-defined]


_ensure_backend_specialization_alias()


@functools.cache
def get_triton_find_paths_if() -> Callable[..., object]:
    if hasattr(triton_jit, "find_paths_if"):
        return triton_jit.find_paths_if
    if hasattr(triton_jit, "_find_paths_if"):
        return triton_jit._find_paths_if  # type: ignore[attr-defined]
    raise AttributeError("Unable to locate Triton find_paths_if helper")


@functools.cache
def get_triton_iterable_path() -> Callable[..., object]:
    if hasattr(triton_jit, "get_iterable_path"):
        return triton_jit.get_iterable_path
    if hasattr(triton_jit, "_get_iterable_path"):
        return triton_jit._get_iterable_path  # type: ignore[attr-defined]
    raise AttributeError("Unable to locate Triton get_iterable_path helper")


def supports_tensor_descriptor() -> bool:
    # call private func we can patch in testing
    return _supports_tensor_descriptor()


@functools.cache
def _supports_tensor_descriptor() -> bool:
    def _cuda_tensor_desc_available() -> bool:
        if not torch.cuda.is_available():
            return False
        major, _ = torch.cuda.get_device_capability(torch.cuda.current_device())
        return major >= 9

    def _xpu_tensor_desc_available() -> bool:
        if not torch.xpu.is_available():
            return False

        from packaging import version

        return version.parse(triton.__version__) >= version.parse("3.5")

    if not (_cuda_tensor_desc_available() or _xpu_tensor_desc_available()):
        return False

    return hasattr(triton.language, "make_tensor_descriptor") or hasattr(
        triton.language, "_experimental_make_tensor_descriptor"
    )


@functools.cache
def get_tensor_descriptor_fn_name() -> str:
    if hasattr(triton.language, "make_tensor_descriptor"):
        return "tl.make_tensor_descriptor"
    assert hasattr(triton.language, "_experimental_make_tensor_descriptor")
    return "tl._experimental_make_tensor_descriptor"


@functools.cache
def torch_dtype_to_tl(torch_dtype: torch.dtype) -> object:
    """Return the `triton.language` dtype that matches a `torch.dtype`."""
    name_str = triton_type(torch_dtype).replace("tl.", "")
    return getattr(tl, name_str)


def min_dot_size(
    device: torch.device, lhs: torch.dtype, rhs: torch.dtype
) -> tuple[int, int, int]:
    # call private func we can patch in testing
    return _min_dot_size(device, lhs, rhs)


@functools.cache
def _min_dot_size(
    device: torch.device, lhs: torch.dtype, rhs: torch.dtype
) -> tuple[int, int, int]:
    if device.type not in ["cuda", "xpu"]:
        # TODO(jansel): support other hardware backends properly besides CUDA and XPU
        return (16, 16, 16)

    if torch.xpu.is_available():
        from triton.backends.intel.compiler import (  # pyright: ignore[reportMissingImports]
            min_dot_size as min_dot_size_xpu,
        )

        device_properties = torch.xpu.get_device_properties()
        gpu_target_info = {
            k: getattr(device_properties, k)
            for k in device_properties.__dir__()
            if not k.startswith("_")
        }

        dot_size_val = min_dot_size_xpu(gpu_target_info)(
            torch_dtype_to_tl(lhs), torch_dtype_to_tl(rhs)
        )
        return tuple(int(v) for v in dot_size_val)  # pyright: ignore[reportReturnType]

    from triton.backends.nvidia.compiler import min_dot_size as min_dot_size_cuda

    props = DeviceProperties.create(device)
    return min_dot_size_cuda(
        GPUTarget(
            backend=props.type,
            arch=props.cc,
            warp_size=props.warp_size or 32,
        )
    )(torch_dtype_to_tl(lhs), torch_dtype_to_tl(rhs))


def warps_to_threads(num_warps: int) -> int:
    if torch.cuda.is_available():
        props = DeviceProperties.create(
            torch.device("cuda", torch.cuda.current_device())
        )
        return num_warps * (props.warp_size or 32)
    return num_warps * 32
