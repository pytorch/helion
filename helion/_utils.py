from __future__ import annotations

import collections
from typing import Sequence

counters: collections.defaultdict[str, collections.Counter[str]] = (
    collections.defaultdict(collections.Counter)
)


def create_shape_matching_slices(
    shape1: Sequence[int], shape2: Sequence[int]
) -> tuple[slice, ...]:
    """Create slices to match the smaller of two shapes.

    This is used for masking tensors to compatible shapes by taking the
    minimum size in each dimension.

    Args:
        shape1: First shape (can be torch.Size or any sequence of ints)
        shape2: Second shape (can be torch.Size or any sequence of ints)

    Returns:
        Tuple of slices that can be used to index a tensor
    """
    return tuple(slice(0, min(d1, d2)) for d1, d2 in zip(shape1, shape2, strict=False))


def convert_size_arg(size: object) -> object:
    """Convert a size argument that may contain RefTile objects.

    Handles:
    - Single RefTile -> int (block_size)
    - List/tuple containing RefTiles -> list with converted sizes
    - Other values -> unchanged
    """
    # Import here to avoid circular dependency
    from helion.language.ref_tile import RefTile

    if isinstance(size, (list, tuple)):
        return [convert_size_arg(item) for item in size]
    if isinstance(size, RefTile):
        return size._block_size
    return size


def convert_tile_indices_to_slices(index: object) -> object:
    """Convert RefTile objects in index to their corresponding slice objects.

    Args:
        index: Index that may contain RefTile objects or tuples of indices

    Returns:
        Index with RefTile objects replaced by their slice objects
    """
    # Import here to avoid circular dependency
    from helion.language.ref_tile import RefTile

    def _extract_slice(obj: object) -> object:
        return obj._slice if isinstance(obj, RefTile) else obj

    if isinstance(index, tuple):
        return tuple(_extract_slice(idx) for idx in index)
    return _extract_slice(index)
