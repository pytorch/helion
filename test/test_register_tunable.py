from __future__ import annotations

import unittest
from unittest.mock import patch

import torch

import helion
from helion import _compat
from helion._testing import DEVICE
from helion._testing import RefEagerTestBase
from helion._testing import TestCase
from helion._testing import code_and_output
from helion._testing import skipIfRocm
from helion.autotuner import EnumFragment
from helion.autotuner import IntegerFragment
from helion.autotuner import PowerOfTwoFragment
import helion.language as hl
from helion.language import loops


class TestRegisterTunable(RefEagerTestBase, TestCase):
    maxDiff = 10000

    def test_power_of_two_fragment_basic(self):
        @helion.kernel(autotune_effort="none")
        def kernel_with_tunable(x: torch.Tensor) -> torch.Tensor:
            (n,) = x.size()
            out = torch.empty_like(x)

            # Register a tunable parameter for block size
            block_size = hl.register_tunable("foo", PowerOfTwoFragment(16, 256))

            for tile_n in hl.tile([n], block_size=[block_size * 2]):
                out[tile_n] = x[tile_n] * 2.0

            return out

        x = torch.randn(128, device=DEVICE, dtype=torch.float32)
        code, result = code_and_output(kernel_with_tunable, (x,))
        expected = x * 2.0
        torch.testing.assert_close(result, expected)
        self.assertIsInstance(
            self.getUserDefinedTunable(
                kernel_with_tunable.bind((x,)).config_spec.user_defined_tunables, "foo"
            ),
            PowerOfTwoFragment,
        )
        self.assertExpectedJournal(code)

    @patch.object(_compat, "_supports_tensor_descriptor", lambda: False)
    @patch.object(loops, "_supports_warp_specialize", lambda: False)
    def test_integer_fragment(self):
        @helion.kernel()
        def kernel_with_int_param(x: torch.Tensor) -> torch.Tensor:
            (n,) = x.size()
            out = torch.empty_like(x)
            # Register an integer tunable parameter
            multiplier = hl.register_tunable("multiplier", IntegerFragment(1, 10, 3))
            for tile_n in hl.tile([n]):
                out[tile_n] = x[tile_n] * multiplier
            return out

        x = torch.randn(128, device=DEVICE, dtype=torch.float32)
        code, result = code_and_output(
            kernel_with_int_param, (x,), block_size=64, multiplier=4
        )
        expected = x * 4
        torch.testing.assert_close(result, expected)
        self.assertExpectedJournal(
            repr(kernel_with_int_param.bind((x,)).config_spec.default_config())
        )
        self.assertExpectedJournal(code)

    def test_enum_fragment(self):
        @helion.kernel(config={"operation": 2})
        def kernel_with_enum(x: torch.Tensor) -> torch.Tensor:
            (n,) = x.size()
            out = torch.empty_like(x)

            # Register an enum tunable parameter
            operation = hl.register_tunable("operation", EnumFragment((1, 2, 4)))

            for tile_n in hl.tile([n], block_size=[64]):
                out[tile_n] = x[tile_n] * operation

            return out

        x = torch.randn(128, device=DEVICE, dtype=torch.float32)
        result = kernel_with_enum(x)
        expected = x * 2.0
        torch.testing.assert_close(result, expected)

    def test_tensor_allocated_with_block_size(self):
        @helion.kernel()
        def fn(x: torch.Tensor):
            m = x.size(0)
            block_m = hl.register_block_size(m)
            tiles_m = (m + block_m - 1) // block_m  # cdiv
            partial = torch.zeros(tiles_m, dtype=x.dtype, device=x.device)
            for tile in hl.tile(m, block_size=block_m):
                partial[tile.begin // block_m] = x[tile].sum()
            return partial.sum()

        x = torch.randn(1024, device=DEVICE, dtype=torch.float32)
        code, result = code_and_output(fn, (x,), block_size=64)
        self.assertExpectedJournal(code)
        torch.testing.assert_close(result, x.sum())

    @patch.object(_compat, "_supports_tensor_descriptor", lambda: False)
    @skipIfRocm("failure on rocm")
    def test_matmul_split_k(self):
        """Test matmul_split_k kernel with register_tunable"""

        @helion.kernel(
            config=helion.Config(
                block_sizes=[32, 64, 64],
                loop_orders=[[1, 2, 0]],
                num_warps=16,
                num_stages=8,
                indexing="block_ptr",
                split_k=64,
            )
        )
        def matmul_split_k(x: torch.Tensor, y: torch.Tensor) -> torch.Tensor:
            m, k = x.size()
            k2, n = y.size()
            assert k == k2, f"size mismatch {k} != {k2}"
            out = torch.zeros(
                [m, n], dtype=torch.promote_types(x.dtype, y.dtype), device=x.device
            )
            split_k = hl.register_tunable("split_k", PowerOfTwoFragment(1, 256))
            k_block = helion.next_power_of_2(helion.cdiv(k, split_k))
            for tile_m, tile_n, outer_k in hl.tile(
                [m, n, k], block_size=[None, None, k_block]
            ):
                acc = hl.zeros([tile_m, tile_n], dtype=torch.float32)
                for inner_k in hl.tile(outer_k.begin, outer_k.end):
                    acc = torch.addmm(acc, x[tile_m, inner_k], y[inner_k, tile_n])
                hl.atomic_add(out, [tile_m, tile_n], acc)
            return out

        m, k, n = 64, 4096, 64
        x = torch.randn([m, k], device=DEVICE, dtype=torch.float16)
        y = torch.randn([k, n], device=DEVICE, dtype=torch.float16)

        code, result = code_and_output(matmul_split_k, (x, y))
        expected = x @ y
        torch.testing.assert_close(result, expected, rtol=1e-2, atol=1)
        self.assertIsInstance(
            self.getUserDefinedTunable(
                matmul_split_k.bind((x, y)).config_spec.user_defined_tunables, "split_k"
            ),
            PowerOfTwoFragment,
        )
        self.assertExpectedJournal(code)


if __name__ == "__main__":
    unittest.main()
