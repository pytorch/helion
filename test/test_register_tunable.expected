This file is automatically generated by assertExpectedJournal calls in test_register_tunable.py.
Update expected outputs by running tests with the EXPECTTEST_ACCEPT=1 environment variable set.

--- assertExpectedJournal(TestRegisterTunable.test_integer_fragment)
helion.Config(block_sizes=[128], range_unroll_factors=[0], range_warp_specializes=[], range_num_stages=[0], range_multi_buffers=[None], range_flattens=[None], num_warps=4, num_stages=3, indexing='pointer', pid_type='flat', multiplier=3)

--- assertExpectedJournal(TestRegisterTunable.test_integer_fragment)
from __future__ import annotations

import torch
import triton
import triton.language as tl
from helion.runtime import default_launcher as _default_launcher

import test.test_register_tunable as _source_module

@triton.jit
def _helion_kernel_with_int_param(x, out, out_stride_0, x_stride_0, n, multiplier, _BLOCK_SIZE_0: tl.constexpr):
    pid_0 = tl.program_id(0)
    offset_0 = pid_0 * _BLOCK_SIZE_0
    indices_0 = (offset_0 + tl.arange(0, _BLOCK_SIZE_0)).to(tl.int32)
    mask_0 = indices_0 < n
    load = tl.load(x + indices_0 * x_stride_0, mask_0, other=0)
    v_0 = multiplier.to(tl.float32)
    v_1 = load * v_0
    tl.store(out + indices_0 * out_stride_0, v_1, mask_0)

def kernel_with_int_param(x: torch.Tensor, *, _launcher=_default_launcher):
    n, = x.size()
    out = torch.empty_like(x)
    multiplier = 4
    _BLOCK_SIZE_0 = 64
    _launcher(_helion_kernel_with_int_param, (triton.cdiv(n, _BLOCK_SIZE_0),), x, out, out.stride(0), x.stride(0), n, multiplier, _BLOCK_SIZE_0, num_warps=4, num_stages=3)
    return out

--- assertExpectedJournal(TestRegisterTunable.test_matmul_split_k)
from __future__ import annotations

import torch
import helion
import triton
import triton.language as tl
from helion.runtime import default_launcher as _default_launcher

import test.test_register_tunable as _source_module

@triton.jit
def _helion_matmul_split_k(x, y, out, out_stride_0, out_stride_1, x_stride_0, x_stride_1, y_stride_0, y_stride_1, n, k, m, _BLOCK_SIZE_1: tl.constexpr, _BLOCK_SIZE_2: tl.constexpr, _BLOCK_SIZE_0: tl.constexpr, _BLOCK_SIZE_3: tl.constexpr):
    num_blocks_0 = tl.cdiv(n, _BLOCK_SIZE_1)
    num_blocks_1 = tl.cdiv(k, _BLOCK_SIZE_2)
    pid_0 = tl.program_id(0) % num_blocks_0
    pid_1 = tl.program_id(0) // num_blocks_0 % num_blocks_1
    pid_2 = tl.program_id(0) // (num_blocks_0 * num_blocks_1)
    offset_1 = pid_0 * _BLOCK_SIZE_1
    indices_1 = (offset_1 + tl.arange(0, _BLOCK_SIZE_1)).to(tl.int32)
    mask_1 = indices_1 < n
    offset_2 = pid_1 * _BLOCK_SIZE_2
    offset_0 = pid_2 * _BLOCK_SIZE_0
    indices_0 = (offset_0 + tl.arange(0, _BLOCK_SIZE_0)).to(tl.int32)
    mask_0 = indices_0 < m
    acc = tl.full([_BLOCK_SIZE_0, _BLOCK_SIZE_1], 0.0, tl.float32)
    tile_end = tl.minimum(offset_2 + _BLOCK_SIZE_2, k)
    for offset_3 in tl.range(offset_2.to(tl.int32), tile_end.to(tl.int32), _BLOCK_SIZE_3):
        indices_3 = offset_3 + tl.arange(0, _BLOCK_SIZE_3).to(tl.int32)
        mask_3 = indices_3 < tile_end
        acc_copy = acc
        acc_copy_0 = acc_copy
        load = tl.load(x + (indices_0[:, None] * x_stride_0 + indices_3[None, :] * x_stride_1), mask_0[:, None] & mask_3[None, :], other=0)
        load_1 = tl.load(y + (indices_3[:, None] * y_stride_0 + indices_1[None, :] * y_stride_1), mask_3[:, None] & mask_1[None, :], other=0)
        acc = tl.dot(load, load_1, acc=acc_copy_0, input_precision='tf32')
    tl.atomic_add(out + (indices_0[:, None] * out_stride_0 + indices_1[None, :] * out_stride_1), acc, mask=mask_0[:, None] & mask_1[None, :], sem='relaxed')

def matmul_split_k(x: torch.Tensor, y: torch.Tensor, *, _launcher=_default_launcher):
    m, k = x.size()
    k2, n = y.size()
    assert k == k2, f'size mismatch {k} != {k2}'
    out = torch.zeros([m, n], dtype=torch.promote_types(x.dtype, y.dtype), device=x.device)
    split_k = 64
    k_block = helion.next_power_of_2(helion.cdiv(k, split_k))
    _BLOCK_SIZE_1 = 64
    _BLOCK_SIZE_2 = k_block
    _BLOCK_SIZE_0 = 32
    _BLOCK_SIZE_3 = 64
    _launcher(_helion_matmul_split_k, (triton.cdiv(n, _BLOCK_SIZE_1) * triton.cdiv(k, _BLOCK_SIZE_2) * triton.cdiv(m, _BLOCK_SIZE_0),), x, y, out, out.stride(0), out.stride(1), x.stride(0), x.stride(1), y.stride(0), y.stride(1), n, k, m, _BLOCK_SIZE_1, _BLOCK_SIZE_2, _BLOCK_SIZE_0, _BLOCK_SIZE_3, num_warps=16, num_stages=8)
    return out

--- assertExpectedJournal(TestRegisterTunable.test_power_of_two_fragment_basic)
from __future__ import annotations

import torch
import triton
import triton.language as tl
from helion.runtime import default_launcher as _default_launcher

import test.test_register_tunable as _source_module

@triton.jit
def _helion_kernel_with_tunable(x, out, out_stride_0, x_stride_0, n, _BLOCK_SIZE_0: tl.constexpr):
    pid_0 = tl.program_id(0)
    offset_0 = pid_0 * _BLOCK_SIZE_0
    indices_0 = (offset_0 + tl.arange(0, _BLOCK_SIZE_0)).to(tl.int32)
    mask_0 = indices_0 < n
    load = tl.load(x + indices_0 * x_stride_0, mask_0, other=0)
    v_0 = 2.0
    v_1 = load * v_0
    tl.store(out + indices_0 * out_stride_0, v_1, mask_0)

def kernel_with_tunable(x: torch.Tensor, *, _launcher=_default_launcher):
    n, = x.size()
    out = torch.empty_like(x)
    block_size = 16
    _BLOCK_SIZE_0 = 2 * block_size
    _launcher(_helion_kernel_with_tunable, (triton.cdiv(n, _BLOCK_SIZE_0),), x, out, out.stride(0), x.stride(0), n, _BLOCK_SIZE_0, num_warps=4, num_stages=3)
    return out

--- assertExpectedJournal(TestRegisterTunable.test_tensor_allocated_with_block_size)
from __future__ import annotations

import torch
import triton
import triton.language as tl
from torch._inductor.runtime import triton_helpers
from helion.runtime import default_launcher as _default_launcher

@triton.jit
def _helion_fn(x, partial, partial_stride_0, x_stride_0, m, _BLOCK_SIZE_0: tl.constexpr):
    pid_0 = tl.program_id(0)
    offset_0 = pid_0 * _BLOCK_SIZE_0
    indices_0 = (offset_0 + tl.arange(0, _BLOCK_SIZE_0)).to(tl.int32)
    mask_0 = indices_0 < m
    load = tl.load(x + indices_0 * x_stride_0, mask_0, other=0)
    sum_1 = tl.sum(load, 0)
    floordiv = triton_helpers.div_floor_integer(offset_0, _BLOCK_SIZE_0)
    tl.store(partial + floordiv * partial_stride_0, sum_1, None)

def fn(x: torch.Tensor, *, _launcher=_default_launcher):
    m = x.size(0)
    block_m = 64
    tiles_m = (m + block_m - 1) // block_m
    partial = torch.zeros(tiles_m, dtype=x.dtype, device=x.device)
    _BLOCK_SIZE_0 = 64
    _launcher(_helion_fn, (triton.cdiv(m, _BLOCK_SIZE_0),), x, partial, partial.stride(0), x.stride(0), m, _BLOCK_SIZE_0, num_warps=4, num_stages=3)
    return partial.sum()
